Add Hugging Face Inference Endpoints as a rest embedder guide (#3053)
meilisearch/documentation: guides/embedders/huggingface.mdx

---
title: Semantic Search with Hugging Face Inference Endpoints - Meilisearch documentation
description: This guide will walk you through the process of setting up Meilisearch with Hugging Face Inference Endpoints.
---

# Semantic search with Hugging Face Inference Endpoints

## Introduction

> This guide will walk you through the process of setting up Meilisearch with [Hugging Face Inference Endpoints](https://ui.endpoints.huggingface.co/) embeddings to enable semantic search capabilities.

Reviewer: ```suggestion
This guide will walk you through the process of setting up a Meilisearch REST embedder with [Hugging Face Inference Endpoints](https://ui.endpoints.huggingface.co/) to enable semantic search capabilities.

<Capsule intent="note" title="`rest` or `huggingface`?">
You can use Hugging Face and Meilisearch in two ways: running the model locally by setting the embedder source to `huggingface`, or remotely in Hugging Face's servers by setting the embeder source to `rest`.
</Capsule>
```

## Requirements

To follow this guide, you'll need:

- A [Meilisearch Cloud](https://www.meilisearch.com/cloud) project running version 1.10 or above with the Vector store activated
- A [Hugging Face account](https://huggingface.co/) with a deployed inference endpoint
- The endpoint URL and API key of the deployed model on your Hugging Face account

## Configure the embedder

Set up an embedder using the update settings endpoint:

```json
{
  "hf-inference": {
    "source": "rest",
    "url": "ENDPOINT_URL",
    "apiKey": "API_KEY",
    "dimensions": 384,
    "documentTemplate": "CUSTOM_LIQUID_TEMPLATE",
    "request": {

>       "inputs": ["{{text}}", "{{..}}"]

Reviewer: Shouldn't we have `model` here? Or outside `request`?

    },
    "response": ["{{embedding}}", "{{..}}"]
  }
}
```

In this configuration:

- `source`: declares Meilisearch should connect to this embedder via its REST API
- `url`: replace `ENDPOINT_URL` with the address of your Hugging Face model endpoint
- `apiKey`: replace `API_KEY` with your Hugging Face API key
- `dimensions`: specifies the dimensions of the embeddings, which are 384 for `baai/bge-small-en-v1.5`
- `documentTemplate`: an optional but recommended [template](/learn/ai_powered_search/getting_started_with_ai_search) for the data you will send the embedder
- `request`: defines the structure and parameters of the request Meilisearch will send to the embedder
- `response`: defines the structure of the embedder's  response

> Once you've configured the embedder, Meilisearch will automatically generate embeddings for your documents. You monitor the task using the Cloud UI or the [get task endpoint](/reference/api/tasks).

Reviewer: ```suggestion
Once you've configured the embedder, Meilisearch will automatically generate embeddings for your documents. Monitor the task using the Cloud UI or the [get task endpoint](/reference/api/tasks).
```

---

> title: Semantic Search with Hugging Face Inference Endpoints - Meilisearch documentation

Reviewer: Why are we focussing on semantic search specifically for this one, instead of going with hybrid?

Author: I've reused the snippet used by similar guides like Cohere, Cloudflare, etc... to keep consistency, we can definitely go with the hybrid instead.

description: This guide will walk you through the process of setting up Meilisearch with Hugging Face Inference Endpoints to enable semantic search capabilities.
---

# Semantic search with Hugging Face Inference Endpoints

## Introduction

This guide will walk you through the process of setting up Meilisearch with [Hugging Face Inference Endpoints](https://ui.endpoints.huggingface.co/) embeddings to enable semantic search capabilities. By leveraging Meilisearch's AI features and Hugging Face Inference Endpoints, you can enhance your search experience and retrieve more relevant results.

## Requirements

To follow this guide, you'll need:

> - A [Meilisearch Cloud](https://www.meilisearch.com/cloud?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide) project running version 1.10 or above with the Vector store activated.

Author: Other guides in that section also uses utm_campaign

> - A [Meilisearch Cloud](https://www.meilisearch.com/cloud?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide) project running version 1.10 or above with the Vector store activated.

Reviewer: Yeap, I'm aware of it. According to Ilia it's not super high priority, so I haven't prioritised removing existing UTMs and am just avoiding adding new ones.

- An Hugging Face account with a deployed inference endpoint. You can sign up for a Hugging Face account at [Hugging Face](https://huggingface.co/).
- The endpoint url and API key of the deployed model on your Hugging Face account.
- No backend required.

> ## Setting up Meilisearch

Reviewer: ```suggestion
## Configure the embedder
```

To set up an embedder in Meilisearch, you need to configure it to your settings. You can refer to the [Meilisearch documentation](/reference/api/settings?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide#update-embedder-settings) for more details on updating the embedder settings.

Hugging Face offers a [wide variety of embedding models](https://ui.endpoints.huggingface.co/catalog?task=sentence-embeddings) you can deploy to embed the content of your documents.

> Here's an example of embedder settings using [BAAI/bge-small-en-v1.5](https://huggingface.co/BAAI/bge-small-en-v1.5) model:

Author: Here the link to the hugging face catalog can be interesting to be kept because it list the category of model Meilisearch can work with (sentence-embeddings)

```json
{
  "hf-inference": {

>     "source": "rest",

Author: Using the `huggingface` source will load the model in local and use the resources. The benefits of using rest is that we can connect to hugging face inference endpoint to offload the resource consumption there and also gain versatility in choosing on which type of resource the model runs

- `apiKey`: Replace `<API Key>` with your actual Hugging Face API key.
- `dimensions`: Specifies the dimensions of the embeddings. Set to 384 for `baai/bge-small-en-v1.5` in our example.
- `documentTemplate`: Optionally, you can provide a [custom template](/learn/ai_powered_search/getting_started_with_ai_search?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide#documenttemplate) for generating embeddings from your documents.
- `request`: Defines the request structure for the deployed model API, including the input parameters.
- `response`: Defines the expected response structure from the deployed model API, including the embedding data.

Once you've configured the embedder settings, Meilisearch will automatically generate embeddings for your documents and store them in the vector store.

It's recommended to monitor the tasks queue to ensure everything is running smoothly. You can access the tasks queue using the Cloud UI or the [Meilisearch API](/reference/api/tasks?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide#get-tasks).

> ## Testing semantic search

Reviewer: ```suggestion
## Perform a semantic search
```

> With the embedder set up, you can now perform semantic searches using Meilisearch. When you send a search query, Meilisearch will generate an embedding for the query using the configured embedder and then use it to find the most semantically similar documents in the vector store.
> To perform a semantic search, you simply need to make a normal search request but include the hybrid parameter:

Reviewer: ```suggestion
With the embedder set up, you can now perform semantic searches. Make a search request with the `hybrid` search parameter, setting `semanticRatio` to `1`:
```

```json

> {
>   "q": "<Query made by the user>",
>   "hybrid": {
>     "semanticRatio": 1,
>     "embedder": "hf-inference"
>   }
> }

Reviewer: ```suggestion
{
  "q": "QUERY_TERMS",
  "hybrid": {
    "semanticRatio": 1,
    "embedder": "hf-inference"
  }
}
```

```

In this request:

> - `q`: Represents the user's search query.
> - `hybrid`: Specifies the configuration for the hybrid search.
>   - `semanticRatio`: Allows you to control the balance between semantic search and traditional search. A value of 1 indicates pure semantic search, while a value of 0 represents full-text search. You can adjust this parameter to achieve a hybrid search experience.
>   - `embedder`: The name of the embedder used for generating embeddings. Make sure to use the same name as specified in the embedder configuration, which in this case is "hf-inference".

Reviewer: ```suggestion
- `q`: the search query
- `hybrid`: enables AI-powered search functionality
  - `semanticRatio`: controls the balance between semantic search and full-text search. Setting it to `1` means you will only receive semantic search results
  - `embedder`: the name of the embedder used for generating embeddings
```

> You can use the Meilisearch API or client libraries to perform searches and retrieve the relevant documents based on semantic similarity.
> 

Reviewer: ```suggestion
```

## Conclusion

> By following this guide, you should now have Meilisearch set up with Hugging Face Inference Endpoints, enabling you to leverage semantic search capabilities in your application. Meilisearch's auto-batching and efficient handling of embeddings make it a powerful choice for integrating semantic search into your project.

Reviewer: ```suggestion
You have set up with an embedder using Hugging Face Inference Endpoints. This allows you to use pure semantic search capabilities in your application.
```

> To explore further configuration options for embedders, consult the [detailed documentation about the embedder setting possibilities](/reference/api/settings?utm_campaign=vector-search&utm_source=docs&utm_medium=huggingface-embeddings-guide#embedders-experimental).

Reviewer: ```suggestion
Consult the [embedder setting documentation](/reference/api/settings#embedders-experimental) for more information on other embedder configuration options.
```
